feat(reminders): AddEditReminderDialog

ericli3690 opened the pull request:
## Purpose / Description
Allows users to add and edit review reminders.

**All commit messages, together:**
- `initializeScheduleRemindersDeckSpinner` doesn't properly handle what happens if showAllDecks is true. This makes pressing on a deck select the wrong deck. This commit fixes this.
- Added TimePicker styling, used when users edit the time of a review reminder.
- The AddEditReminderDialog displays the current time in the TimePicker when a review reminder is being created. Adds a utility function to ReviewReminderTime's companion object to calculate this time. I decided to put this function within ReviewReminderTime so it can be accessed by both AddEditReminderDialog and AddEditReminderDialogViewModel.
- XML file for AddEditReminderDialog. Allows time and deck modification. Also has a dropdown for advanced settings, including minimum card trigger threshold (with more advanced settings hopefully coming soon). The advanced settings will be grouped under a dropdown to prevent them from overwhelming users when the dialog initially opens.
- Created a ViewModel for the AddEditReminderDialog. A specific view model is created because this allows the dialog's state to persist across redraws, ex. if the theme changes, if the device rotates, etc. It also centralizes the data of the review reminder being edited in a single source of truth.
- Creates AddEditReminderDialog.
- Adds DialogMode, which represents whether the dialog is in adding or editing mode.
- Sets submit, cancel, and delete actions for the dialog; this is the main way users can delete review reminders. Deletion is locked behind a confirmation box just in case the user accidentally clicks the button.
- Results for the deck picker dropdown in the dialog are received in ScheduleReminders and sent to the AddEditReminderDialog via a FragmentResult. See the docstring of `onDeckSelected` in ScheduleReminders to see why this is done.
- Filling in the values in the dialog fields and setting listeners for when the values change is pulled out into separate methods for readability.
- Adds `showTimePickerDialog`. Shows a TimePicker dialog for picking review reminder time via modern Material3 guidelines. Overrides `onConfigurationChanged` to properly handle device rotation, which for some reason is not handled by default by MaterialTimePicker.
- An edited review reminder is passed back to ReviewRemindersDatabase as a FragmentResult, which is a simple way of passing information between fragments.
- Adds a fragment result listener to ScheduleReminders to detect when an AddEditReminderDialog has completed, and if so, edits the database and UI accordingly.
- Filled out the `addReminder` and `editReminder` methods to show the AddEditReminderDialog.

**UI Screenshots:**
Dialog:
<img width="835" height="880" alt="image" src="https://github.com/user-attachments/assets/e11896ff-276b-4e9b-864e-fd218beb60d6" />
<img width="831" height="879" alt="image" src="https://github.com/user-attachments/assets/8b30cd70-f490-4243-a4c0-0086002ab5be" />

Error handling:
<img width="1092" height="780" alt="image" src="https://github.com/user-attachments/assets/a3d4ee5d-37a2-4245-b48b-b92024d9e602" />

Sub-dialogs:
<img width="830" height="873" alt="image" src="https://github.com/user-attachments/assets/eca30c9a-6a6a-4eaa-b298-d6b4113af6c7" />

TimePickers:
<img width="817" height="873" alt="image" src="https://github.com/user-attachments/assets/05d1f743-e214-4e72-91f1-ee71ba0000c6" />
<img width="825" height="881" alt="image" src="https://github.com/user-attachments/assets/118ab9e5-57be-47a5-86aa-01e4e230b93e" />

Tablet:
<img width="1497" height="903" alt="image" src="https://github.com/user-attachments/assets/5b236e90-5a11-41ed-9119-979d073431ae" />
<img width="1492" height="899" alt="image" src="https://github.com/user-attachments/assets/0e12ddb5-0764-4ecb-b91a-c60bdfc6943e" />

## Fixes
For GSoC 2025: Review Reminders

## Approach
Creates a ViewModel to store the state of the dialog. Information is passed between the dialog and ScheduleReminders via fragment results, which is a simple and straightforward API.

## How Has This Been Tested?
- Works on a physical Samsung S23, API 34; can create, read, update, and delete review reminders. I've been casually testing this code for the last month as I've developed other features, and I think it should be fairly stable.
- Works on a Medium Tablet, API 36.
- I plan on writing instrumented tests in the last week of August.

## Learning (optional, can help others)
- This code has been waiting to be merged for a while! Really happy to see it finally up as a PR.
- Had a lot of trouble in the final stretch with TimePicker. Turns out there are three different ways to make it: using XML, using the provided TimePickerDialog, and using MaterialTimePicker. The latter is the only way to get the nice styling. Also, the background was always tinted a bit blue no matter what I did. Turns out I had to add `<item name="elevationOverlayEnabled">false</item>`.
- Also had trouble making sure the TimePicker dialog redrew whenever you rotated your screen. I ended up redrawing it manually since there didn't seem to be built-in handling for it.

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

<!--- Uncomment this section ONLY if this PR introduces new resources (external libraries, icons etc)
## Licenses
_For each new external resource, add a row to the table below:_

| Library | Description | License |
| --- | --- | --- |
| Sample Icon Library | Sample Description | [The Apache Software License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0.txt) |

**Maintainers:**

* [ ] Add the https://github.com/ankidroid/Anki-Android/labels/Licenses label
* [ ] Update the [licenses](https://github.com/ankidroid/Anki-Android/wiki/Licences) wiki when merging
--->

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/AddEditReminderDialogViewModel.kt:
```diff
@@ -0,0 +1,73 @@
+        _deckSelected.value = deckId
+    }
+
+    fun setCardTriggerThreshold(threshold: Int?) {
```
Fixed.

ericli3690 commented:
Resolved Copilot feedback. Implemented a Google Accessibility Scanner recommendation.

criticalAY commented on AnkiDroid/src/main/res/values/styles.xml:
```diff
@@ -254,4 +254,11 @@
         <item name="android:navigationBarColor" tools:ignore="NewApi">@android:color/transparent</item>
     </style>
+
+    <style name="TimePickerStyle" parent="ThemeOverlay.Material3.MaterialTimePicker">
+        <item name="elevationOverlayEnabled">false</item>
+        <item name="colorTertiaryContainer">?colorPrimary</item>
+        <item name="colorOnTertiaryContainer">?colorOnPrimary</item>
+        <item name="colorPrimaryContainer">?colorPrimary</item>
+    </style>
```
👍🏼

criticalAY commented on AnkiDroid/src/main/res/layout/add_edit_reminder_dialog.xml:
```diff
@@ -0,0 +1,190 @@
+                        android:orientation="horizontal">
+
+                        <!-- The following image's icon switches between ic_baseline_chevron_right_24 and ic_expand_more_black_24dp_xml -->
+
```
Can remove this line? or maybe comment too or is it needed?

ericli3690 replied:
@criticalAY Like Arthur, I'm also a little confused. I think what you mean is that I should remove this comment? I've gone ahead and done that. Let me know if you wanted something else instead.

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/AddEditReminderDialogViewModel.kt:
```diff
@@ -0,0 +1,73 @@
+ * the input type the user is using to enter the threshold into the app. In other words, this class reflects the concrete
+ * EditText fields in the dialog, not abstract backend data representations.
+ */
+class AddEditReminderDialogViewModel(
```
Nice!

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/AddEditReminderDialog.kt:
```diff
@@ -0,0 +1,412 @@
+        }
+
+        // 2. state -> viewModel: Configure how the UI reacts to changes in the state
+        viewModel.time.observe(this) { time ->
```
Why are we not using `viewLifecycleOwner`?

criticalAY replied:
```
viewModel.time.observe(viewLifecycleOwner) { time -> … }
viewModel.advancedSettingsOpen.observe(viewLifecycleOwner) { advancedSettingsOpen -> … }

```
Can we do better here, inspect?

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/AddEditReminderDialog.kt:
```diff
@@ -0,0 +1,412 @@
+        }
+
+        // Fill in initial field values
+        timeButton.text = viewModel.time.value.toString()
```
rely only on observers?

ericli3690 replied:
You're right, that line is unnecessary. Deleted.

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/AddEditReminderDialog.kt:
```diff
@@ -0,0 +1,412 @@
+        return dialog
+    }
+
+    fun setUpDialogFields() {
```
split this method into smaller ones

ericli3690 replied:
Done!

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt:
```diff
@@ -113,6 +116,23 @@ class ScheduleReminders :
+        // If the user creates or edits a review reminder, the dialog for doing so opens
+        // Once their changes are complete, the dialog closes and this fragment is reloaded
+        // Hence, we check for any fragment results here and update the database accordingly
+        setFragmentResultListener(ADD_EDIT_DIALOG_RESULT_REQUEST_KEY) { _, bundle ->
+            val modeOfFinishedDialog =
+                BundleCompat.getParcelable(
+                    requireArguments(),
+                    ACTIVE_DIALOG_MODE_ARGUMENTS_KEY,
+                    AddEditReminderDialog.DialogMode::class.java,
+                )
+            val newOrModifiedReminder = BundleCompat.getParcelable(bundle, ADD_EDIT_DIALOG_RESULT_REQUEST_KEY, ReviewReminder::class.java)
+
+            Timber.d("Dialog result received with recent dialog mode: %s", modeOfFinishedDialog)
+            if (modeOfFinishedDialog == null) return@setFragmentResultListener
+            handleAddEditDialogResult(newOrModifiedReminder, modeOfFinishedDialog)
+        }
```
Too dense extract it out

ericli3690 replied:
I've cleaned up the code here a bit but I don't think it can be extracted in any meaningful way. Here's what it looks like now:

```
        setFragmentResultListener(ADD_EDIT_DIALOG_RESULT_REQUEST_KEY) { _, bundle ->
            val modeOfFinishedDialog =
                BundleCompat.getParcelable(
                    requireArguments(),
                    ACTIVE_DIALOG_MODE_ARGUMENTS_KEY,
                    AddEditReminderDialog.DialogMode::class.java,
                ) ?: return@setFragmentResultListener
            val newOrModifiedReminder =
                BundleCompat.getParcelable(
                    bundle,
                    ADD_EDIT_DIALOG_RESULT_REQUEST_KEY,
                    ReviewReminder::class.java,
                )
            Timber.d("Dialog result received with recent dialog mode: %s", modeOfFinishedDialog)
            handleAddEditDialogResult(newOrModifiedReminder, modeOfFinishedDialog)
        }
```

Is there a specific way you were thinking of extracting out some of this code?

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt:
```diff
@@ -145,6 +165,86 @@ class ScheduleReminders :
+    /**
+     * When a [AddEditReminderDialog] instance finishes, we handle the result of the dialog fragment via this method.
+     */
+    private fun handleAddEditDialogResult(
```
too much in one method, attaching a patch

criticalAY replied:
```patch
Subject: [PATCH] feat: add browsing preference category
---
Index: AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt b/AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt	(revision c16568065039a8952cbccbc21ab4256db6675a81)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt	(date 1755642278336)
@@ -173,76 +173,69 @@
         modeOfFinishedDialog: AddEditReminderDialog.DialogMode,
     ) {
         Timber.d("Handling add/edit dialog result: mode=%s reminder=%s", modeOfFinishedDialog, newOrModifiedReminder)
-        // Modify database
+
+        // Database updates
         launchCatchingTask {
             catchDatabaseExceptions {
                 if (modeOfFinishedDialog is AddEditReminderDialog.DialogMode.Edit) {
-                    // Delete the existing reminder if we're in edit mode
-                    // This action must be separated from writing the modified reminder because the user may have updated the reminder's deck,
-                    // meaning we need to delete the old reminder in the old deck, then add a new reminder to the new deck
                     val reminderToDelete = modeOfFinishedDialog.reminderToBeEdited
                     Timber.d("Deleting old reminder from database")
-                    val deleteReminderWithId: (
-                        HashMap<ReviewReminderId, ReviewReminder>,
-                    ) -> Map<ReviewReminderId, ReviewReminder> = { reminders ->
-                        reminders.remove(reminderToDelete.id)
-                        reminders
-                    }
                     when (reminderToDelete.scope) {
-                        is ReviewReminderScope.Global -> database.editAllAppWideReminders(deleteReminderWithId)
-                        is ReviewReminderScope.DeckSpecific ->
-                            database.editRemindersForDeck(
-                                reminderToDelete.scope.did,
-                                deleteReminderWithId,
-                            )
+                        is ReviewReminderScope.Global -> database.editAllAppWideReminders(deleteReminder(reminderToDelete))
+                        is ReviewReminderScope.DeckSpecific -> database.editRemindersForDeck(reminderToDelete.scope.did, deleteReminder(reminderToDelete))
                     }
                 }
 
-                newOrModifiedReminder?.let {
+                newOrModifiedReminder?.let { reminder ->
                     Timber.d("Writing new or modified reminder to database")
-                    // There is a new / updated reminder that must be written to the database
-                    val writeNewOrModifiedReminder: (
-                        HashMap<ReviewReminderId, ReviewReminder>,
-                    ) -> Map<ReviewReminderId, ReviewReminder> = { reminders ->
-                        reminders[it.id] = it
-                        reminders
-                    }
-                    when (it.scope) {
-                        is ReviewReminderScope.Global -> database.editAllAppWideReminders(writeNewOrModifiedReminder)
-                        is ReviewReminderScope.DeckSpecific ->
-                            database.editRemindersForDeck(
-                                it.scope.did,
-                                writeNewOrModifiedReminder,
-                            )
-                    }
-                }
-            }
+                    when (reminder.scope) {
+                        is ReviewReminderScope.Global -> database.editAllAppWideReminders(upsertReminder(reminder))
+                        is ReviewReminderScope.DeckSpecific -> database.editRemindersForDeck(reminder.scope.did, upsertReminder(reminder))
+                    }
+                }
+            }
+        }
+
+        // UI updates
+        updateUi(modeOfFinishedDialog, newOrModifiedReminder)
+
+        // Feedback
+        showSnackbar(
+            when (modeOfFinishedDialog) {
+                is AddEditReminderDialog.DialogMode.Add -> "Successfully added new review reminder"
+                is AddEditReminderDialog.DialogMode.Edit -> if (newOrModifiedReminder == null) {
+                    "Successfully deleted review reminder"
+                } else {
+                    "Successfully edited review reminder"
+                }
+            }
+        )
+    }
+
+    private fun deleteReminder(reminder: ReviewReminder) =
+        { reminders: HashMap<ReviewReminderId, ReviewReminder> ->
+            reminders.remove(reminder.id)
+            reminders
+        }
+
+    private fun upsertReminder(reminder: ReviewReminder) =
+        { reminders: HashMap<ReviewReminderId, ReviewReminder> ->
+            reminders[reminder.id] = reminder
+            reminders
         }
 
-        // Modify UI
-        if (modeOfFinishedDialog is AddEditReminderDialog.DialogMode.Edit) {
+    private fun updateUi(mode: AddEditReminderDialog.DialogMode, newReminder: ReviewReminder?) {
+        if (mode is AddEditReminderDialog.DialogMode.Edit) {
             Timber.d("Deleting old reminder from UI")
-            reminders.remove(modeOfFinishedDialog.reminderToBeEdited.id)
+            reminders.remove(mode.reminderToBeEdited.id)
         }
-        newOrModifiedReminder?.let {
+        newReminder?.let {
             if (scheduleRemindersScope == ReviewReminderScope.Global || scheduleRemindersScope == it.scope) {
                 Timber.d("Adding new reminder to UI")
                 reminders[it.id] = it
             }
         }
         triggerUIUpdate()
-
-        showSnackbar(
-            when (modeOfFinishedDialog) {
-                is AddEditReminderDialog.DialogMode.Add -> "Successfully added new review reminder"
-                is AddEditReminderDialog.DialogMode.Edit -> {
-                    when (newOrModifiedReminder) {
-                        null -> "Successfully deleted review reminder"
-                        else -> "Successfully edited review reminder"
-                    }
-                }
-            },
-        )
     }
 
     /**

```

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt:
```diff
@@ -145,6 +165,86 @@ class ScheduleReminders :
+        }
+        triggerUIUpdate()
+
+        showSnackbar(
```
Use an anchor to avoid this
<img width="374" height="110" alt="Screenshot 2025-08-20 at 3 56 47 AM" src="https://github.com/user-attachments/assets/46142b18-4f6d-4d1e-88a2-b16b2c02140a" />

Arthur-Milchior reviewed (commented):
I tried to review, but very little to state on top of what @criticalAY mentioned

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt:
```diff
@@ -256,6 +266,30 @@ class ScheduleReminders :
         const val DECK_SELECTION_RESULT_REQUEST_KEY = "reminder_deck_selection_result_request_key"
 
+        private const val SERIALIZATION_ERROR_MESSAGE =
+            "Something went wrong. A serialization error was encountered while working with review reminders."
```
It's hard-coded because I'm not sure if its wording is in its final form, i.e. I don't want it translated yet. Thus, I've added a TODO comment. Thanks!

ericli3690 commented:
Minor edits:
- Non-functional change moving view retrieval from the `onCreateDialog` method to the individual sub-methods for better readability.
- Added more information to a ViewModel logger.

ericli3690 commented:
Resolved merge conflict.

ericli3690 commented:
- Resolved an issue caused by one of David's recent changes related to SelectedDeck. Unfortunately, until DeckSpinnerSelection is cleaned up, we'll still need to use the constant it provides in the AddEditReminderDialog.
- Rebased.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/AddEditReminderDialog.kt:
```diff
@@ -0,0 +1,418 @@
+        // dismiss conditionally from within the click listener methods (see onSubmit and onDelete).
+        dialog.setOnShowListener {
+            val positiveButton = dialog.getButton(AlertDialog.BUTTON_POSITIVE)
+            val negativeButton = dialog.getButton(AlertDialog.BUTTON_NEGATIVE)
```
Huh, you're right. I'll double check that.

ericli3690 commented:
Thanks for the review! I'll take a look at implementing your suggested changes shortly.

ericli3690 commented:
- Rebased
- Adapted to use the updated `toFormattedString` method introduced at https://github.com/ankidroid/Anki-Android/pull/19265
- Added a small NF commit to remove a useless lint suppression annotation I found

ericli3690 commented:
@david-allison Thanks for the fast merge on the 24h-time-formatting PR I created! Do you think we could quickly get this guy in, too? I'll be able to put more code up for review once it's in, and I think I've addressed all your feedback from the last time you looked at it. I've heard you're on vacation, though; if that's the case, please ignore this!
